src/Adapters/Auth/index.js

```diff
@@ -59,21 +61,63 @@ const providers = {
   microsoft,
   keycloak,
   ldap,
+  webauthn,
 };
 
-function authDataValidator(adapter, appIds, options) {
-  return function (authData) {
-    return adapter.validateAuthData(authData, options).then(() => {
-      if (appIds) {
-        return adapter.validateAppId(appIds, authData, options);
+function authDataValidator(provider, adapter, appIds, options) {
+  return async function (authData, req, user) {
+    if (appIds && typeof adapter.validateAppId === 'function') {
+      await adapter.validateAppId(appIds, authData, options, req, user);
+    }
+    if (typeof adapter.validateAuthData === 'function') {
+      return adapter.validateAuthData(authData, options, req, user);
+    } else if (
+      typeof adapter.validateSetUp === 'function' &&
+      typeof adapter.validateLogin === 'function' &&
+      typeof adapter.validateUpdate === 'function'
+    ) {
+      // We can consider for DX purpose when masterKey is detected, we should
+      // trigger a logged in user
+      const isLoggedIn =
+        (req.auth.user && user && req.auth.user.id === user.id) || (user && req.auth.isMaster);
+      let hasAuthDataConfigured = false;
+
+      if (user && user.get('authData') && user.get('authData')[provider]) {
+        hasAuthDataConfigured = true;
+      }
+
+      if (isLoggedIn) {
+        // User is currently updating his authData
+        if (hasAuthDataConfigured) {
+          return adapter.validateUpdate(authData, options, req, user);
+        }
+        // Let's setup if the user does not have the provider configured
+        return adapter.validateSetUp(authData, options, req, user);
+      }
+
+      // Not logged in and authData configured into the DB
```

Comment:
"configured into the DB" -> what does that mean? Maybe "authData is stored in DB"?